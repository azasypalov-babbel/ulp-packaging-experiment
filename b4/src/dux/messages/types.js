const NAMESPACE = 'MESSAGES';

export const ADD_MESSAGE = `${NAMESPACE}/ADD_MESSAGE`;
export const REMOVE_MESSAGE = `${NAMESPACE}/REMOVE_MESSAGE`;
