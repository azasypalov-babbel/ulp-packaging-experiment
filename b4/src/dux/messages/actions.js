import * as types from './types';

export const addMessage = (name) => ({
  type: types.ADD_MESSAGE,
  payload: name
});

export const removeMessage = (name) => ({
  type: types.REMOVE_MESSAGE,
  payload: name
});
