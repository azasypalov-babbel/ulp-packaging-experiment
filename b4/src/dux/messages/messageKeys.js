export const MESSAGE_KEYS = {
  MIC_GENERIC: 'mic-generic',
  MIC_PERMISSIONS: 'mic-permissions',
  MIC_MUTED: 'mic-muted',
  NETWORK: 'network',
  GENERIC: 'generic'
};
