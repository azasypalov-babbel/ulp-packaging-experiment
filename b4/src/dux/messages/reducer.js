import * as types from './types';

const initialState = [];

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case types.ADD_MESSAGE: {
      const name = action.payload;

      if (state.includes(name)) {
        return state;
      }

      return [
        ...state,
        name
      ];
    }

    case types.REMOVE_MESSAGE: {
      return state.filter((name) => name !== action.payload);
    }

    default: {
      return state;
    }
  }
}
