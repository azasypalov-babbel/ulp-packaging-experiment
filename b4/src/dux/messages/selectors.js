
export const getFirstMessage = (state) => state[0];

