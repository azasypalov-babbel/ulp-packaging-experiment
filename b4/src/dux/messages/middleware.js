import * as permissionsTypes from '../permissions/types';
import { addMessage, removeMessage } from './actions';
import { PERMISSIONS_STATUS } from '../permissions/status';
import { isMicEnabled } from '../session/selectors';
import { MESSAGE_KEYS } from './messageKeys';

const messages = (store) => (next) => (action) => {
  const result = next(action);

  const isMicPermissionAction = [
    permissionsTypes.UPDATE_MIC_PERMISSION,
    `${permissionsTypes.REQUEST_MIC_PERMISSIONS}_FULFILLED`
  ].includes(action.type);

  if (isMicPermissionAction) {
    const { session, permissions } = store.getState();
    const micPermission = action.payload || permissions.micPermission;

    const permissionsGranted = [
      PERMISSIONS_STATUS.prompt,
      PERMISSIONS_STATUS.granted
    ].includes(micPermission);

    const trainerRequiresMicPermission = isMicEnabled(session);

    if (!permissionsGranted && trainerRequiresMicPermission) {
      store.dispatch(addMessage(MESSAGE_KEYS.MIC_PERMISSIONS));
    } else {
      store.dispatch(removeMessage(MESSAGE_KEYS.MIC_PERMISSIONS));
    }
  }

  return result;
};

export default messages;
