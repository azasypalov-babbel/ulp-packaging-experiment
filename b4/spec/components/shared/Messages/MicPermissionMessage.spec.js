import React from 'react';
import { shallow } from 'enzyme';
import { MicPermissionMessage } from
  '../../../../src/components/shared/Messages/MicPermissionMessage';
import { MESSAGE_KEYS } from '../../../../src/dux/messages/messageKeys';


describe('<MicPermissionMessage />', () => {
  let wrapper;
  const mockTrack = jest.fn();

  const defaultProps = {
    setMicSettings: jest.fn(),
    removeMessage: jest.fn(),
    locale: 'en',
    learnLanguageAlpha3: 'DEU',
    track: mockTrack,
    isReview: false,
    translations: {
      title: 'speech_recognition.error_mic_permission.title',
      body: 'speech_recognition.error_mic_permission.body',
      cta: 'speech_recognition.error_mic_permission.cta'
    }
  };

  test('tracks gui:shown', () => {
    shallow(<MicPermissionMessage {...defaultProps} />);

    expect(mockTrack).toHaveBeenCalledTimes(1);
    expect(mockTrack.mock.calls[0][0]).toMatchSnapshot();
    mockTrack.mockClear();
  });

  describe('when in lesson', () => {
    test('renders with the button', () => {
      const props = { ...defaultProps, isReview: false };
      wrapper = shallow(<MicPermissionMessage {...props} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('when in review', () => {
    test('renders without the button', () => {
      const props = { ...defaultProps, isReview: true };
      wrapper = shallow(<MicPermissionMessage {...props} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('on click', () => {
    beforeEach(() => {
      wrapper = shallow(<MicPermissionMessage {...defaultProps} />);
      const ctaButton = wrapper.find('[data-selector="cta-button"]');
      ctaButton.simulate('click');
    });

    afterEach(() => {
      mockTrack.mockClear();
    });

    test('calls setMicSettings', () => {
      expect(defaultProps.setMicSettings).toHaveBeenCalledWith(false);
    });

    test('calls removeMessage', () => {
      expect(defaultProps.removeMessage).toHaveBeenCalledWith(MESSAGE_KEYS.MIC_PERMISSIONS);
    });

    test('tracks gui:interacted', () => {
      expect(mockTrack).toHaveBeenCalledTimes(2);
      // first time will be called on mount for track gui:shown
      expect(mockTrack.mock.calls[1][0]).toMatchSnapshot();
    });
  });
});
