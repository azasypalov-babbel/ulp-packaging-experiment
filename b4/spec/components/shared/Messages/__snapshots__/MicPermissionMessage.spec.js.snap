// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MicPermissionMessage /> on click tracks gui:interacted 1`] = `
Object {
  "event": "gui:interacted",
  "payload": Object {
    "gui_container": "mic_error_no_permission",
    "gui_element": "opt_out_button",
    "interaction": "clicked",
    "learn_language_alpha3": "DEU",
    "locale": "en",
    "origin": "lesson_player",
  },
  "version": 1,
}
`;

exports[`<MicPermissionMessage /> tracks gui:shown 1`] = `
Object {
  "event": "gui:shown",
  "payload": Object {
    "gui_container": "mic_error_no_permission",
    "gui_element": "opt_out_button",
    "learn_language_alpha3": "DEU",
    "locale": "en",
    "origin": "lesson_player",
  },
  "version": 1,
}
`;

exports[`<MicPermissionMessage /> when in lesson renders with the button 1`] = `
<ErrorOverlay
  body={
    <ReplaceInline
      replacement={
        <VideocamBlocked
          size="1.25rem"
        />
      }
      symbol="[*]"
    >
      speech_recognition.error_mic_permission.body
    </ReplaceInline>
  }
  dataSelector="error-overlay-mic-permission-error"
  icon={
    <WithTheme(MicOff)
      size="3.5rem"
    />
  }
  key="permission-error"
  title="speech_recognition.error_mic_permission.title"
>
  <WithKeypress(styled.button)
    data-selector="cta-button"
    onClick={[Function]}
  >
    speech_recognition.error_mic_permission.cta
  </WithKeypress(styled.button)>
</ErrorOverlay>
`;

exports[`<MicPermissionMessage /> when in review renders without the button 1`] = `
<ErrorOverlay
  body={
    <ReplaceInline
      replacement={
        <VideocamBlocked
          size="1.25rem"
        />
      }
      symbol="[*]"
    >
      speech_recognition.error_mic_permission.body
    </ReplaceInline>
  }
  dataSelector="error-overlay-mic-permission-error"
  icon={
    <WithTheme(MicOff)
      size="3.5rem"
    />
  }
  key="permission-error"
  title="speech_recognition.error_mic_permission.title"
/>
`;
